### ISSUE TITLE
How to detect dev vs production

### ISSUE THREAD

**Issue by kristojorg** [zeit/next.js#1365]
Forgive me if this is an extremely basic question, but I'm struggling to figure out how to determine if my code is running in dev or production. 

I tried adding env vars to my `scripts` in `package.json`: 
```
  "scripts": {
    "dev": "ENV=dev next",
    "build": "ENV=production next build",
    "start": "ENV=production next start"
  }
```
though I don't know how to access these variables. I tried:
```
const ENV = 'undefined' !== typeof window ? window.ENV : process.ENV
````
but this didn't seem to work. These variables don't show up in `window` object from browser. 

From my understanding, `window` is available on client and `process` is available on server, but I'm not sure how to set a var on them, or access that var appropriately. Alternatively, is there an existing variable that indicates production vs dev mode?

Thanks!

**Comment by timneutkens** [comment]
Please have a look at https://github.com/zeit/next.js/tree/master/examples/with-universal-configuration 😄

**Comment by sedubois** [comment]
I suggest to use NODE_ENV.

**Comment by arunoda** [comment]
Thanks guys.
I think this is answered.

**Comment by kristojorg** [comment]
Thank you all!

**Comment by firrae** [comment]
Hi there,

I don't mean to necro this but I'm having an issue when trying to work this into my project. I only see the root error when I clear my cache which is this image:
<img width="1920" alt="screen shot 2017-06-22 at 09 39 47" src="https://user-images.githubusercontent.com/6454839/27437002-ed933882-572e-11e7-9538-ca247d76a5f2.png">

That follows the documentation on it and makes sense, but the error doesn't to me. I have copied the example verbatim into my project, even into the same locations, so I'm confused as to what is missing as when I run the isolated example it seems to work? I started working off the MobX example so maybe it's something in there?

```javascript
const prod = process.env.NODE_ENV === 'production'

module.exports = {
  'BACKEND_URL': prod ? 'https://api.example.com' : 'https://localhost:8080'
}
```

Any help is appreciated, thanks,

Steve

**Comment by raelmiu** [comment]
Can't get the example working, and the many layers of interacting libs in the example makes it hard to follow.

1. I've installed babel-plugin-transform-define
2. I've added an env-config.js
3. I've added exporting variables in the env-config
4. When I run the project, my process.enc.X variables are undefined. This is true both when I run it as DEV and build->start

I'm using a server to render my pages, so not running through "next up" or whatever the command is.

**Comment by johntran** [comment]
@raelmiu 

Have you tried https://github.com/zeit/next.js#exposing-configuration-to-the-server--client-side

**Comment by onomatopoetry** [comment]
How could you use the [article](https://github.com/zeit/next.js#exposing-configuration-to-the-server--client-side) mentioned by @johntran to set up two sets of configuration - one for dev and one for prod?

My app communicates with an api, often in the context of a page's `getInitialProps()` using [isomorphic-unfetch](https://www.npmjs.com/package/isomorphic-unfetch). When pages are rendered server-side fetch requires an absolute url, so I've set up my node.config.js like so:

```
module.exports = {
  publicRuntimeConfig: {
    API_ENDPOINT: 'http://localhost:3000/api'
  }
}
```

But this endpoint is the development version of the api, I can't have my app contacting the dev api in production! How can I change the API endpoint in node.config.js when the code is running in my production environment?

**Comment by johntran** [comment]
@onomatopoetry 
```
const prodConfig = {
  publicRuntimeConfig: {
    API_ENDPOINT: 'http://google.com/api'
  }
}

const devConfig = {
  publicRuntimeConfig: {
    API_ENDPOINT: 'http://localhost:3000/api'
  }
}

module.exports = process.env.NODE_ENV === 'production ? prodConfig : devConfig
```

**Comment by malixsys** [comment]
In case someone (else) comes here through Google, check [this example]...(https://github.com/zeit/next.js/blob/canary/examples/with-env-from-next-config-js/next.config.js)

**Comment by rafaelcalhau** [comment]
> Please have a look at https://github.com/zeit/next.js/tree/master/examples/with-universal-configuration 😄

Page is not available

**Comment by Asrover** [comment]
My full example with custom webpack config:

```
const {
    PHASE_DEVELOPMENT_SERVER,
    PHASE_PRODUCTION_BUILD,
} = require('next/constants')

module.exports = (phase) => {
    env: (() => {
      // when started in development mode `next dev` or `npm run dev` regardless of the value of STAGING environmental variable
      const isDev = phase === PHASE_DEVELOPMENT_SERVER
      // when `next build` or `npm run build` is used
      const isProd = phase === PHASE_PRODUCTION_BUILD && process.env.STAGING !== '1'
      // when `next build` or `npm run build` is used
      const isStaging =
        phase === PHASE_PRODUCTION_BUILD && process.env.STAGING === '1'

      console.log(`isDev:${isDev}  isProd:${isProd}   isStaging:${isStaging}`)

      return {
        RESTURL_SPEAKERS: (() => {
          if (isDev) return 'http://localhost:4000/speakers'
          if (isProd) {
            return 'https://www.siliconvalley-codecamp.com/rest/speakers/ps'
          }
          if (isStaging) return 'http://localhost:11639'
          return 'RESTURL_SPEAKERS:not (isDev,isProd && !isStaging,isProd && isStaging)'
        })(),
        RESTURL_SESSIONS: (() => {
          if (isDev) return 'http://localhost:4000/sessions'
          if (isProd) return 'https://www.siliconvalley-codecamp.com/rest/sessions'
          if (isStaging) return 'http://localhost:11639'
          return 'RESTURL_SESSIONS:not (isDev,isProd && !isStaging,isProd && isStaging)'
        })(),
      }
    })(),
    webpack: (config) => {
        // Some webpack config
        return config
    },
}
```

**Comment by schmidsi** [comment]
A bit hacky but a working solution is this:

```js
// next.config.js
module.exports = {
  publicRuntimeConfig: {
    IS_LOCAL_DEV: process.argv.includes('dev')
  }
}
```

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
